ironfish-rust/src/spending.rs

```diff
@@ -62,132 +44,138 @@ pub struct SpendParams {
-            proof_generation_key: Some(proof_generation_key),
-            payment_address: Some(note.owner.sapling_payment_address()),
-            auth_path: sapling_auth_path(witness),
-            commitment_randomness: Some(note.randomness),
-            anchor: Some(witness.root_hash()),
+        let circuit = Spend {
+            value_commitment: Some(self.value_commitment.clone()),
+            proof_generation_key: Some(spender_key.sapling_proof_generation_key()),
+            payment_address: Some(self.note.owner.sapling_payment_address()),
+            auth_path: self.auth_path.clone(),
+            commitment_randomness: Some(self.note.randomness),
+            anchor: Some(self.root_hash),
             ar: Some(public_key_randomness),
         };
-        let proof = groth16::create_random_proof(spend_circuit, &SAPLING.spend_params, &mut OsRng)?;
 
+        // Proof that the spend was valid and successful for the provided owner
+        // and note.
+        let proof =
+            groth16::create_random_proof(circuit, &SAPLING.spend_params, &mut thread_rng())?;
+
+        // The public key after randomization has been applied. This is used
+        // during signature verification. Referred to as `rk` in the literature
+        // Calculated from the authorizing key and the public_key_randomness.
         let randomized_public_key = redjubjub::PublicKey(spender_key.authorizing_key.into())
             .randomize(public_key_randomness, SPENDING_KEY_GENERATOR);
-        let nullifier = note.nullifier(&spender_key, witness_position(witness));
 
-        Ok(SpendParams {
-            spender_key,
-            public_key_randomness,
+        // Bytes to be placed into the nullifier set to verify whether this note
+        // has been previously spent.
+        let nullifier = self.note.nullifier(spender_key, self.witness_position);
+
+        let blank_sig = {
+            let buf = [0u8; 64];
+            Signature::read(&mut buf.as_ref())?
+        };
+
+        let spend_proof = SpendProof {
             proof,
-            value_commitment,
+            value_commitment: value_commitment_point,
             randomized_public_key,
-            root_hash: witness.root_hash(),
-            tree_size: witness.tree_size(),
+            root_hash: self.root_hash,
+            tree_size: self.tree_size,
             nullifier,
+            authorizing_signature: blank_sig,
+        };
+
+        Ok(UnsignedSpendProof {
+            public_key_randomness,
+            spend_proof,
         })
     }
+}
 
-    /// Sign this spend with the stored private key, and return a SpendProof
-    /// suitable for serialization.
-    ///
-    /// Verifies the proof before returning to prevent posting broken
-    /// transactions
-    pub fn post(&self, signature_hash: &[u8; 32]) -> Result<SpendProof, IronfishError> {
-        let private_key = redjubjub::PrivateKey(self.spender_key.spend_authorizing_key);
+pub struct UnsignedSpendProof {
```

Comment:
This sort of defines the in-between state that exists during the transaction building. Building the transaction requires the proofs to be generated, and signing the spends requires the transaction to be built. It feels cleaner than mashing it all in one step as it is now. This is ultimately what lets us remove a lot of the duplicated code between the spend/output params and spend/output proofs.